sno/merge.py

```diff
@@ -187,34 +183,121 @@ def handle_merge_conflicts(
             _get_pk_for_index_entry(*x) for x in zip(datasets3, index_entries3)
         )
         if "META" in pks3:
+            click.echo(f"Merge conflict found in metadata for {dataset_path}")
             raise NotYetImplemented(
                 "Sorry, resolving conflicts in metadata isn't supported yet"
             )
         pk = first_true(pks3)
         if pks3.count(pk) != 3:
+            click.echo(
+                f"Merge conflict found where primary keys have changed in {dataset_path}"
+            )
             for i in range(3):
-                print(f"{labels3[i]}: {dataset_path}:{pks3[i]}")
+                click.echo(f"{labels3[i]}: {dataset_path}:{pks3[i]}")
             raise NotYetImplemented(
                 "Sorry, resolving conflicts where primary keys have changed isn't supported yet"
             )
         conflict_pks.setdefault(dataset_path, [])
         conflict_pks[dataset_path].append(pk)
 
-    prefixes3 = ("- ", "+ ", "+ ")
-    big_prefixes3 = ("---", "+++", "+++")
-    cols3 = ("red", "green", "green")
+    num_conflicts = sum(len(pk_list) for pk_list in conflict_pks.values())
+    click.echo(f"\nFound {num_conflicts} conflicting features:")
+    for dataset_path, pks in conflict_pks.items():
+        click.echo(f"{len(pks)} in {dataset_path}")
+    click.echo()
+
+    # Check for dirty working copy before continuing - we don't want to fail after interactive part.
+    ours_rs = repo_structures3[OURS]
+    ours_rs.working_copy.reset(commits3[OURS], ours_rs, force=force)
+
+    # At this point, the failure should be dealt with so we can start resolving conflicts interactively.
+    # We don't want to fail during conflict resolution, since then we would lose all the user's work.
+    # TODO: Support other way(s) of resolving conflicts.
+    if sys.stdout.isatty():
+        _interactive_pause(
+            "Press enter to begin resolving merge conflicts, or Ctrl+C to abort at any time..."
+        )
+    else:
+        click.echo(
+            "Printing conflicts but not resolving - merge conflicts must be resolved in an interactive terminal"
```

Comment:
We need to test this... somehow.

Maybe we should add a `-X {ours|theirs|ancestor}` option like `git merge` has, then we can use it to specify the behaviour in testing. That sidesteps the need to test terminal input.

Is `ancestor` ever a valid choice? It makes sense to display it I think, because it can be useful to understand how the two branches have diverged from their starting position, but I can't imagine a reason for ever choosing it to resolve a conflict.

Reply:
> Maybe we should add a -X {ours|theirs|ancestor} option like git merge has, then we can use it to specify the behaviour in testing. That sidesteps the need to test terminal input.

👍 

> Is ancestor ever a valid choice? It makes sense to display it I think, because it can be useful to understand how the two branches have diverged from their starting position, but I can't imagine a reason for ever choosing it to resolve a conflict.

"Make both these things go away and I'll fix it myself"?